File: Step2/Step2_Data_Analysis.f90
program step2_data_analysis_fortran
    implicit none
    character(len=2048) :: line, part
    integer :: status, i, j, k, l, pipe_count
    real :: charge
    integer :: behavior_val
    real, dimension(100) :: unique_charges
    integer, dimension(100) :: trial_counts, success_counts
    integer :: n_unique = 0
    real :: critical_charge = -1.0
    
    unique_charges = 0.0; trial_counts = 0; success_counts = 0

    open(10, file='Step2/data_repo/observation_data', status='old', action='read')

    do
        read(10, '(A)', iostat=status) line
        if (status /= 0) exit
        
        if (line(1:1) == '|') then
            pipe_count = 0
            do i = 1, len_trim(line)
                if (line(i:i) == '|') then
                    pipe_count = pipe_count + 1
                    
                    if (pipe_count == 6) then
                        j = i + 1
                        do k = j, len_trim(line)
                            if (line(k:k) == '|') exit
                        end do
                        part = adjustl(line(j:k-1))
                        read(part, *, iostat=status) charge
                        
                        if (status == 0) then
                            j = k + 1
                            do l = j, len_trim(line)
                                if (line(l:l) == '|') exit
                            end do
                            part = line(j:l-1)
                            
                            behavior_val = 0
                            if (index(part, 'droop') > 0 .or. index(part, 'close') > 0) then
                                behavior_val = 1
                            end if
                            
                            do j = 1, n_unique
                                if (abs(unique_charges(j) - charge) < 1e-6) then
                                    trial_counts(j) = trial_counts(j) + 1
                                    success_counts(j) = success_counts(j) + behavior_val
                                    exit
                                end if
                            end do
                            if (j > n_unique) then
                                n_unique = n_unique + 1
                                unique_charges(n_unique) = charge
                                trial_counts(n_unique) = 1
                                success_counts(n_unique) = behavior_val
                            end if
                        end if
                        exit 
                    end if
                end if
            end do
        end if
    end do
    close(10)

    do j = 1, n_unique
        if (trial_counts(j) > 0) then
            if (real(success_counts(j))/real(trial_counts(j)) >= 0.5) then
                if (critical_charge < 0.0 .or. unique_charges(j) < critical_charge) then
                    critical_charge = unique_charges(j)
                end if
            end if
        end if
    end do

    open(20, file='Step2/plant_data_analysis_fortran.json', status='replace')
    write(20, '(A)') '{'
    write(20, '(A,F7.4,A)') '    "critical_charge_uC": ', critical_charge, ','
    write(20, '(A)') '    "analysis_summary": ['
    do j = 1, n_unique
        write(20, '(A)') '        {'
        write(20, '(A,F10.6,A)') '            "Charge (uC)": ', unique_charges(j), ','
        write(20, '(A,I4,A)')    '            "Trials": ', trial_counts(j), ','
        write(20, '(A,F7.4)')    '            "Success Rate": ', real(success_counts(j))/real(trial_counts(j))
        if (j < n_unique) then
            write(20, '(A)') '        },'
        else
            write(20, '(A)') '        }'
        end if
    end do
    write(20, '(A)') '    ]'
    write(20, '(A)') '}'
    close(20)

end program step2_data_analysis_fortran

File: Step2/Step2_Data_Analysis.py
import pandas as pd
import numpy as np
import re

# ============================================
# Step 2: オジギソウ実験データの解析 (U: 相転移閾値の抽出)
# データソース: AAA-2003/Electrophysiology-of-Mimosa-pudica-L
# ============================================

def parse_observation_data(file_path):
    with open(file_path, 'r') as f:
        lines = f.readlines()
    
    data = []
    # ヘッダー行をスキップし、パイプ区切りのデータを抽出
    for line in lines:
        if line.startswith('|') and 'Sl.no' not in line:
            cols = [c.strip() for c in line.split('|')]
            if len(cols) >= 8:
                try:
                    # 電荷量 (Net Charge Given) の数値を抽出 (例: "4230μC" -> 4230)
                    charge_str = cols[6]
                    charge_val = float(re.findall(r"[-+]?\d*\.\d+|\d+", charge_str.replace(' ', ''))[0])
                    
                    # 行動 (Behavior) の有無を判定
                    behavior_str = cols[7].lower()
                    success = 1 if ('droop' in behavior_str or 'close' in behavior_str) else 0
                    
                    data.append({'charge_uC': charge_val, 'success': success})
                except:
                    continue
    
    return pd.DataFrame(data)

# データの読み込み
df = parse_observation_data('Step2/data_repo/observation_data')

# 閾値解析
# チャージ量ごとに成功率を計算
summary = df.groupby('charge_uC')['success'].agg(['count', 'mean']).reset_index()
summary.columns = ['Charge (uC)', 'Trials', 'Success Rate']

print("--- Mimosa pudica Phase Transition (U) Analysis ---")
print(summary)

# 臨界電荷量の推定 (成功率が0.5を超える最小のチャージ量)
critical_charge = summary[summary['Success Rate'] >= 0.5]['Charge (uC)'].min()
print(f"\nEstimated Critical Charge (Threshold for U): {critical_charge} uC")

# JSON結果の保存
import json
results = {
    "data_source": "AAA-2003/Electrophysiology-of-Mimosa-pudica-L",
    "critical_charge_uC": float(critical_charge),
    "analysis_summary": summary.to_dict(orient='records')
}
with open('Step2/plant_data_analysis.json', 'w') as f:
    json.dump(results, f, indent=4)

# --- グラフの描画 ---
import matplotlib.pyplot as plt

plt.figure(figsize=(10, 6))
plt.semilogx(summary['Charge (uC)'], summary['Success Rate'], 'o-', color='green', label='Success Rate (Real Data)')
plt.axvline(critical_charge, color='red', linestyle='--', label=f'Threshold (~{critical_charge} uC)')
plt.fill_betweenx([0, 1], 0, critical_charge, color='gray', alpha=0.2, label='No Response Zone')
plt.title("Mimosa pudica: Charge vs Phase Transition Success Rate")
plt.xlabel("Net Charge Given (uC) - Log Scale")
plt.ylabel("Success Rate (Phase Transition to Drooping)")
plt.ylim(-0.05, 1.05)
plt.grid(True, which="both", ls="-", alpha=0.5)
plt.legend()
plt.tight_layout()
plt.savefig("Step2/plant_data_plot.png")

print("\nAnalysis complete. Results saved to Step2/plant_data_analysis.json and plant_data_plot.png")
